### PR TITLE
Improve performance circle to polygon method

### PR DESCRIPTION
**Pull request by nshumoogum**
### What

Updated "CircleToPolygon" method to use go routines to handle calculation of coordinates

### How to review

Check changes make sense

### Who can review

Anyone

### CODE REVIEW

File: geo/geopoint.go [hunk 1]
```diff
@@ -10,16 +10,35 @@ import (
...
-var defaultMaxSegments = maxSegments{segments: 180}
+// Default geo configuration for methods on config receiver
+var DefaultConfig = &Config{
+	defaultConcurrencyLimit: defaultConcurrency,
+	defaultMaxSegments:      defaultMaximumSegments,
+}
+
+// List of errors
+var (
+	ErrRadiusLargerThanEarth    = errors.New("radius can not be greater than the radius of the Earth, 6378137 metres")
```
Comment (cookel2) [inline geo/geopoint.go]: Line 33 seems a little extreme! Are DP using this repo to select areas across the whole Earth or just the UK? If just the UK then maybe half the length of the country should be the maximum radius. According to Google the length of the UK is 965km https://www.nationsencyclopedia.com/Europe/United-Kingdom-LOCATION-SIZE-AND-EXTENT.html. So I guess half is about 482.5km or 482500 metres.

Comment (nshumoogum) [reply geo/geopoint.go]: Remember this is a library open to the public, so anyone could be using it across the globe

Comment (nshumoogum) [reply geo/geopoint.go]: Following from the above, one may think it silly to generate a circular boundary anywhere on the globe that is exceedingly large but how would we know that there isn't a user case for this.

Fyi - I mostly copied this code by converting the same functions from another language into Go for our use (I couldn't find an existing Go library)

File: geo/geopoint.go [hunk 2]
```diff
@@ -33,20 +52,19 @@ type Coordinate struct {
...
-var (
-	ErrRadiusLargerThanEarth    = errors.New("radius can not be greater than the radius of the Earth, 6378137 metres")
-	ErrTooFewSegments           = errors.New("too few segments, this should be set to 3 or more")
-	ErrInvalidLongitudinalPoint = errors.New("longitude has to be between -180 and 180")
-	ErrInvalidLatitudinalPoint  = errors.New("latitude has to be between -90 and 90")
-)
+// New instantiates a geo object with defined values to be used by geo methods
+// e.g. concurrency limit and maximum number of segments when generating a circle from a point
+func New(concurrencyLimit, maximumSegments int) (geo *Config) {
+	geo = DefaultConfig
```
Comment (RMPall) [inline geo/geopoint.go]: (minor): I guess you would have to check the concurrencylimit and maximumSegments here before creating a valid geo config with these parameters ?

File: geo/geopoint.go [hunk 3]
```diff
@@ -57,25 +75,43 @@ func NewGeoPoint(lat, lon float64) (*Coordinate, error) {
...
-		segment := (twoPi * float64(-i)) / float64(segments)
-		coordinate := generateCoordinate(*geoPoint, radius, segment)
-		coordinates = append(coordinates, coordinate)
+
+		semaphoreChan <- struct{}{}
+
+		wg.Add(1)
+
+		go func(i int) {
+			defer func() {
```
Comment (RMPall) [inline geo/geopoint.go]: (suggestion): you don't need the defer here as you are not taking a different path elsewhere in the go routine. you could do this step as the last step inside the goroutine like 
```
sector := (twoPi * float64(-i)) / float64(segments)
coordinate := generateCoordinate(geoPoint, radius, sector)
coordinates[i] = coordinate
<-semaphoreChan // read to release a slot
wg.Done()
```

File: geo/geopoint.go [hunk 4]
```diff
@@ -57,25 +75,43 @@ func NewGeoPoint(lat, lon float64) (*Coordinate, error) {
...
+	var semaphoreChan = make(chan struct{}, geo.defaultConcurrencyLimit)
+
+	var wg sync.WaitGroup
 
 	for i := 0; i < segments; i++ {
-		segment := (twoPi * float64(-i)) / float64(segments)
-		coordinate := generateCoordinate(*geoPoint, radius, segment)
-		coordinates = append(coordinates, coordinate)
+
+		semaphoreChan <- struct{}{}
```
Comment (RMPall) [inline geo/geopoint.go:45]: (question): Not sure if I get the logic right here, you have a semaphoreChan buffered channel with defaultConcurrencyLimit. But then you are already spinning go routines which equals the number of segments. Now are you implying that even though you have started n number of goroutines, the number of goroutines which will be  working in parallel to generate coordinate will be equal to the default concurrency limit ? if so we could start only that many number of goroutines in the first place right, why do we have more goroutines started ?

**Review by RMPall — CHANGES_REQUESTED** [review]

File: geo/geopoint.go (hunk 2 above)
Comment (nshumoogum) [reply geo/geopoint.go]: So I don't know what a valid upper limit for concurrent processes would be, so allowing the user to define this themselves. Im happy to hear other possible solutions.

File: geo/geopoint.go (hunk 4 above)
Comment (nshumoogum) [reply geo/geopoint.go:45]: Yes you are right, I remember thinking this too, but never refactored it. I will use the lowest value between the `geo.defaultConcurrencyLimit` and the number of `segments` passed into method when buffering the number of go routines

**Comment by nshumoogum** [comment]
All the above refactoring was to help the increasing large single test file "geopoint_test.go" and wanting to write more unit tests - hopefully the refactor makes it easier to navigate and will allow us to extend this library for other geographical shapes

File: geo/geopoint.go (hunk 1 above)
Comment (cookel2) [reply geo/geopoint.go]: To be honest I'm not sure who will be using this library so if there's a user case then fair enough.

**Review by RMPall — APPROVED** [review]

**Pull request merged**